README
 BASHLYK - среда управления сценариями командной оболочки BASH от разработки и
 до повторного использования кода.
 Включает в себя средства:
 - прозрачное журналирование
 - контроль запуска (например, защита от повторного запуска)
 - прозрачное манипулирование временными объектами (задачи, файлы, каталоги)
 - манипулирование файлами настроек (чтение/создание/запись), в том числе, в
   стиле ini-конфигураций
 - разбор аргументов командной строки


ChangeLog
Mon, 08 May 2023 13:00:01 +0000
	Fix bugs
	Version 1.96
- fix lib{err,log,pid} tests on Oracle Linux 7.x
- add setup.sh for self installing
libcfg.sh:
- add support for BASH 5.1 and above - "BASH_REMATCH is no longer readonly"
libpid.sh:
- add PID_MAX support
- fix pid::stop tests
- removed unused global variables


Fri, 23 Aug 2019 19:22:32 +0400
	Fix bugs
	Version 1.95
- achieve compatibility with CentOS 7, install with make now works. Needed
  manual preinstall all required packages (ex docbook5-style-xsl.noarch )
- used 'declare -f' instead 'declare -pf' for
- CFG::bind.cli - fixed typos and handling long opt like --a-b
- CFG::load:
  * fixed error of losing the config name
  * replaced [-1] indice for compatibility
- CFG::keys - used linefeed to separate keys
- added CFG::keys.each - functional way for key/value handling
- std::trim - simplified code
- std::isHash - fixed test
- added a net::ipv4.wildcard for Cisco wildcards
src/Makefile.am:
- 'syslib' renamed to 'lib'
- fixed install/uninstall by automake (partially)
- xsl to manpages - added support for RedHat family
- use DESTDIR for hooks
doc/Makefile.am:
- removed generated timestamp from doc/bashlyk.html
debian:
- infrastructure updated, 'debian' folder moved to root of sources
- postinst - removed unused code, tests no longer run during installation
- fixed lintian warnings
- control - to 'Build-Depends' added 'patch'
README.md:
- added asciinema demo
- updated example for liberr.sh
- updated codeblocks
- added launchpad.net repo link


Thu, 06 Sep 2018 03:03:44 +0400
	Fix bugs
	Version 1.94
- improved handling of both command line options and configuration files:
  - CFG::bind.cli - fixed handling of arguments with whitespaces.
  - CFG::storage.use - changed handling of the short filenames, its search is
    first performed in the initialized configuration paths, only then in the
    current directory (this behavior moved from CFG::load). Configuration file
    is checked for reading and the ability to create it in the absence. If the
    specified file can't be used, an appropriate warning is issued and an
    internal configuration file is generated.
  - CFG::load - fixed a processing of the configurations with initial and/or end
    dots in the filename.
- pid::onExit.unlink.empty, pid::trap, std::temp - fixed handling of arguments
  with whitespaces
- added the pid::onExit.unlink.empty function for removing initialized but not
  used files.
- fixed handling of such input lines as '-E', '-e' by using 'printf' instead of
  'echo'
- msg::notify - fixed output logging for non interactive mode
- msg::getXsessionProperties - added partial support for the (X)wayland session
- a limitation was found in using the 'error' command. This command generates
  the code with 'eval', under certain conditions the command interpreter can
  execute the second and subsequent lines of a multi-line argument (for example,
  output of some command). To avoid this, you must apply one of the following
  rules:
  - do not use double quotes in the argument string of the 'error' command or at
    least around variables that may contain multi lines.
  - preliminarily delete linefeeds symbols, for example, with std::inline
- added new error states - AlreadyExist, NoSuchFile, NoSuchDir, TimeExpired and
  error handlers for err::postfix - AlreadyExist, NoSuchFile, NoSuchDir
- added a base support for interactive input in the command interpreter
- improved tests, descriptions and 'makefile' instructions
- libold.sh, libini.sh, ini2csv.awk, inisection2csv.awk are now obsolete and are
  not installed
--details--
libcfg.sh:
- prefix 'config' in the name of the CFG instance sets the INI style
- CFG::get - return error state for empty value
- CFG::set - improved arguments handling
- added helper method - CFG::storage.use.default
- CFG::{read,load,save} - set storage if needed
- CFG::bind.cli:
  - handle a error of the missing args
  - std::whitespace.{en,de}code used for handle the arguments with whitespaces
libstd.sh:
- std::whitespace.decode - improved tests
- added function std::inline for multi lines conversion to single line
bashlyk:
- added global array _bashlyk_aArg for command line options storing
- libini.sh is deprecated
- added external binary list
- check /proc/mounts
debian/control:
- added libc-bin to Depends list


Sun, 15 Apr 2018 14:24:16 +0400
	New upstream release
	Version 1.93
- libini library has been renamed to libcfg, which emphasizes the versatility of
  this library as a means of managing configurations or complex data
  conglomerations.
- Innovations:
  - added a separate method for specifying an external configuration store.
  - you can specify the configuration type (shellmode or INI) by the prefix of
    the instance name of the class CFG.
  - added bSaveConfig option for the instance settings
- liberr - the syntax of the error handling function is changed:
  'on error <action> <code> <hint>' -> 'error <code> <action> <hint>'
  - added the handlers NotNumber, NotExistNotCreated
  - added function err::debugf for formatted output
- fixed debian/postinst bug for a first installation
- updated lists of the external commands and exports
--details--
libcfg.sh:
- CFG::storage.use without argument make $_bashlyk_pathDat directory and
  default storage
- if instance name begin with 'conf' then used "active configuration"
  mode (shellmode), wherein if the key is not a valid variable, then the
  'key = value' string turns into a comment
- added bSaveConfig option for management a preserving configuration on
  the CFG::save method call
- improved testunit for CFG::settings method
- added test for autoswitch testing
- updated descriptions
- changed CFG::{save,read,load,bind.cli} methods for use CFG::storage.use
  method
libmsg.sh:
- msg::notifyTool - fixed variable declaration
bashlyk:
- added global variable _bashlyk_pathCfg - _bashlyk_path{Cnf,Ini} sister
get-external-commands-and-export-list.sh:

src/liblog.sh
#
# $Git: liblog.sh 1.94-44-934 2019-11-29 23:18:16+04:00 yds $
#
#****h* BASHLYK/liblog
#  DESCRIPTION
#    The library contains a set of functions for controlling the output of the
#    script messages
#  USES
#    libstd libmsg
#  AUTHOR
#    Damir Sh. Yakupov <yds@bk.ru>
#******
#***iV* liblog/BASH compatibility
#  DESCRIPTION
#    Compatibility checked by bashlyk (BASH version 4.xx or more required)
#    $_BASHLYK_LIBLOG provides protection against re-using of this module
#  SOURCE
[ -n "$_BASHLYK_LIBLOG" ] && return 0 || _BASHLYK_LIBLOG=1
[ -n "$_BASHLYK" ] || . ${_bashlyk_pathLib}/bashlyk || eval '                  \
                                                                               \
    echo "[!] bashlyk loader required for ${0}, abort.."; exit 255             \
                                                                               \
'
#******
#****L* liblog/Used libraries
#  DESCRIPTION
#    Loading external libraries
#  SOURCE
[[ -s ${_bashlyk_pathLib}/liberr.sh ]] && . "${_bashlyk_pathLib}/liberr.sh"
[[ -s ${_bashlyk_pathLib}/libstd.sh ]] && . "${_bashlyk_pathLib}/libstd.sh"
#******
#****G* liblog/Global variables
#  DESCRIPTION
#    Global variables of the library
#  SOURCE
#: ${_bashlyk_fnLogSock:=}
#: ${_bashlyk_pidLogSock:=}
: ${HOSTNAME:=localhost}
: ${_bashlyk_emailSubj:="${_bashlyk_sUser}@${HOSTNAME}::${_bashlyk_s0}"}
: ${_bashlyk_fnLog:="${_bashlyk_pathLog}/${_bashlyk_s0}.log"}
: ${_bashlyk_bUseSyslog:=0}
: ${_bashlyk_bNotUseLog:=1}
: ${_bashlyk_sCond4Log:=redirect}
: ${_bashlyk_iPidMax:=999999}

declare -rg _bashlyk_aRequiredCmd_log="

    [ date hostname logger mkdir mkfifo rm touch tty

"

declare -rg _bashlyk_aExport_log="

    log::{add,file,ger,init,interactivity,necessity,terminality}

"
#******
#****f* liblog/log::ger
#  SYNOPSIS
#    log::ger <text>
#  DESCRIPTION
#    log engine for adding <text> to log file with standart stamps if logging is
#    setted
#  INPUTS
#    <text> - input text
#  OUTPUT
#    There are four possibilities:
#     * stdout only
#     * $_bashlyk_fnLog only
#     * syslog by logger and stdout
#     * syslog by logger and $_bashlyk_fnLog
#  EXAMPLE
#    local bInteract bNotUseLog bTerminal
#    _ =bInteract
#    _ =bNotUseLog
#    _ =bTerminal
#    local b=true fnExec reT reP s
#    fnExec=$(mktemp --suffix=.sh || tempfile -s .test.sh)
#    reT='[ADFJMNOS][abceglnoprtuyv]{2} [0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}'
#    reP="[[:space:]]$HOSTNAME ${0##*/}\[[[:digit:]]{${#_bashlyk_iPidMax}}\]:[[:space:]].*"
#    cat <<'EOF' > $fnExec                                                      #-
#    local fnLog=$(mktemp --suffix=.log || tempfile -s .test.log)               #-
#    _ fnLog $fnLog                                                             #-
#    _ bInteract 0                                                              #-
#    _ bNotUseLog 0                                                             #-
#    _ bTerminal 0                                                              #-
#    log::init                                                                  #-
#    _ fnLog                                                                    #-
#    log::ger test                                                              #-
#    date                                                                       #-
#    echo $_bashlyk_pidLogSock                                                  #-
#    echo '%d'                                                                  #-
#    EOF                                                                        #-
#    . $fnExec
#    kill $_bashlyk_pidLogSock
#    rm -f $_bashlyk_fnLogSock
#    sleep 0.1
#    while read -t9 s; do [[ $s =~ ^${reT}${reP}$ ]] || b=false; done < $fnLog  #-
#    [[ $b == true ]]                                                           #? true
#    rm -f $fnExec $fnLog
#    _ bInteract "$bInteract"
#    _ bNotUseLog "$bNotUseLog"
#    _ bTerminal "$bTerminal"
#    ## TODO test syslog mode
#  SOURCE
log::ger() {

  local bSysLog bUseLog sTagLog IFS=$' \t\n'

  bSysLog=0
  bUseLog=0

  sTagLog="${_bashlyk_s0}[$(printf -- "%0${#_bashlyk_iPidMax}d" $$)]"

  if [[ -z "$_bashlyk_bUseSyslog" || "$_bashlyk_bUseSyslog" -eq 0 ]]; then

    bSysLog=0

  else

    bSysLog=1

  fi

  if [[ $_bashlyk_bNotUseLog ]]; then

    (( $_bashlyk_bNotUseLog != 0 )) && bUseLog=0 || bUseLog=1

  else

    log::necessity && bUseLog=1 || bUseLog=0

  fi

  mkdir -p "$_bashlyk_pathLog" || error NotExistNotCreated throw $_bashlyk_pathLog

  pid::onExit.unlink $_bashlyk_pathLog

  case "${bSysLog}${bUseLog}" in

    00)
        printf -- '%s\n' "$@"
     ;;

    01)
        log::stamp "$HOSTNAME $sTagLog: $*" >> $_bashlyk_fnLog
     ;;

    10)
        printf -- '%s\n' "$*"
        logger -t "$sTagLog" "$*"
     ;;

    11)
        log::stamp "$HOSTNAME $sTagLog: $*" >> $_bashlyk_fnLog
        logger -t "$sTagLog" "$*"
     ;;

  esac

}
#******
## TODO improve description
#****f* liblog/log::add
#  SYNOPSIS
#    log::add [-] [<text>]
#  DESCRIPTION
#    Wrapper around log::ger to support stream from standard input
#  INPUTS
#    -      - data is expected from standard input
#    <text> - String (tag) for output.
#             If there is a "-" as the first argument, then the string is
#             considered a prefix (tag) for each line from the standard input.
#  OUTPUT
#   Depends on output parameters
#  EXAMPLE
#    # TODO improve test
#    echo -n . | log::add -                                                     | {{ '^\.$'       }}
#    echo test | log::add - tag                                                 | {{ '^tag test$' }}
#  SOURCE
log::add() {

  local sTag s

  if [[ "$1" == "-" ]]; then

    shift
    [[ $* ]] && sTag="$* "

    while read s || [[ $s ]]; do [[ $s ]] && log::ger "${sTag}${s}"; done

  else

    [[ $* ]] && log::ger "$*"

  fi

}
#******
#****f* liblog/log::interactivity
#  SYNOPSIS
#    log::interactivity
#  DESCRIPTION
#    Checking the operating mode of standard input and output devices
#  RETURN VALUE
#    0 - "non-interactive" mode, there is redirection of standard input and/or
#         output
#    1 - "interactive" mode, redirection of standard input and/or output is not
#        detected
#  EXAMPLE
#    log::interactivity                                                         #? true
#    log::interactivity                                                         #= false
#  SOURCE
log::interactivity() {

  [[ -t 1 && -t 0 && $TERM && "$TERM" != "dumb" ]] \
    && _bashlyk_bInteract=1 || _bashlyk_bInteract=0

  return $_bashlyk_bInteract

}
#******
#****f* liblog/log::terminality
#  SYNOPSIS
#    log::terminality
#  DESCRIPTION
#    Checking the presence of a control terminal
#  RETURN VALUE
#    0 - terminal not detected
#    1 - terminal detected
#  EXAMPLE
#    log::terminality                                                           #? false
#    log::terminality                                                           #= false
#  SOURCE
log::terminality() {

  tty > /dev/null 2>&1 && _bashlyk_bTerminal=1 || _bashlyk_bTerminal=0
  return $_bashlyk_bTerminal

}
#******
#****f* liblog/log::necessity
#  SYNOPSIS
#    log::necessity
#  DESCRIPTION
#    Check the conditions for using the log file
#  RETURN VALUE
#    0 - save stdout and stderr to log file
#    1 - logging do not required
#  EXAMPLE
#    _bashlyk_sCond4Log='redirect'
#    log::necessity                                                             #? true
#    log::necessity                                                             #= false
#  SOURCE
log::necessity() {

  log::terminality
  log::interactivity

  case ${_bashlyk_sCond4Log} in

    redirect)
              _bashlyk_bNotUseLog=$_bashlyk_bInteract ;;
      noterm)
              _bashlyk_bNotUseLog=$_bashlyk_bTerminal ;;
           *)
              _bashlyk_bNotUseLog=$_bashlyk_bInteract ;;
  esac

  return $_bashlyk_bNotUseLog

}
#******
#****f* liblog/log::init
#  SYNOPSIS
#    log::init
#  DESCRIPTION
#    Creating a named pipe for redirecting the output of stdout and stderr to a
#     log file with automatic addition of standard stamps.
#  ERRORS
#     1                  - The socket is not created, but the output of the
#                          stdout and the stderr is redirected to the log file
#                          (without stamps)
#     NotExistNotCreated - The socket directory does not exist and can not be
#                          created
#  EXAMPLE
#    local fnLog=$(mktemp --suffix=.log || tempfile -s .test.log)
#    _ fnLog $fnLog                                                             #? true
#    log::init                                                                  #? true
#    ls -l $fnLog                                                               #? true
#    rm -f $fnLog
#  SOURCE
log::init() {

  local fnSock IFS=$' \t\n'

  if [[ $_bashlyk_sArg ]]; then

    fnSock="$(std::getMD5 ${_bashlyk_s0} ${_bashlyk_sArg}).${$}.socket"
    fnSock="${_bashlyk_pathRun}/${fnSock}"

  else

    fnSock="${_bashlyk_pathRun}/${_bashlyk_s0}.${$}.socket"

  fi

  mkdir -p $_bashlyk_pathRun || error NotExistNotCreated return+warn $_bashlyk_pathRun

  [[ -a "$fnSock" ]] && rm -f $fnSock

  if mkfifo -m 0600 $fnSock >/dev/null 2>&1; then

    ( log::add - < $fnSock )&
    _bashlyk_pidLogSock=$!
    exec >>$fnSock 2>&1

    _bashlyk_fnLogSock=$fnSock
    pid::onExit.unlink $fnSock

    return 0

  else

    error NotExistNotCreated warn Socket $fnSock not created..

    exec >>$_bashlyk_fnLog 2>&1

    _bashlyk_fnLogSock=$_bashlyk_fnLog

    return 1

  fi

}
#******
#****f* liblog/log::file
#  SYNOPSIS
#    log::file [<filename>]
#  DESCRIPTION
#    Wrapper around log::init to activate output redirection to the log
#    <filename> with error handling
#  ERRORS
#     NotExistNotCreated - log file can not be created, abort
#  EXAMPLE
#    local fnLog=$(mktemp --suffix=.log || tempfile -s .test.log)
#    rm -f $fnLog
#    log::file $fnLog                                                           #? true
#    ls -l $fnLog                                                               #? true
#    rm -f $fnLog
#  SOURCE
log::file() {

  local IFS=$' \t\n'

  case "$1" in
          '') ;;
    ${1##*/}) _bashlyk_fnLog="${_bashlyk_pathLog}/$1";;
           *)
              _bashlyk_fnLog="$1"
              _bashlyk_pathLog=${_bashlyk_fnLog%/*}
           ;;
  esac

  {

    mkdir -p "$_bashlyk_pathLog"
    touch "$_bashlyk_fnLog"

  } || error NotExistNotCreated throw $_bashlyk_fnLog

  log::init

  return 0

}
#******
#****f* liblog/log::stamp
#  SYNOPSIS
#    log::stamp <text>
#  DESCRIPTION
#    Show input <text> with time stamp in format 'Mar 28 10:03:40' (LC_ALL=C)
#  INPUTS
#    <text> - suffix to the header
#  OUTPUT
#    input <text> with time stamp in format 'Mar 28 10:03:40' (LC_ALL=C)
#  EXAMPLE
#    local re
#    re='^[ADFJMNOS][abceglnoprtuyv]{2} [0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2} AB$'
#    log::stamp AB | {{ -E "$re" }}
#  SOURCE

if (( _bashlyk_ShellVersion > 4002000 )); then

  log::stamp() { LC_ALL=C printf -- '%(%b %d %H:%M:%S)T %s\n' '-1' "$*"; }

else

  log::stamp() { LC_ALL=C date "+%b %d %H:%M:%S ${*//%/%%}"; }

fi
#******
